### ISSUE TITLE
improve remap coercion story

### ISSUE THREAD

**Issue by JeanMertz** [timberio/vector#5960]
TL;DR:

1. rename all type coercion `to_*` functions by removing the `to_` part (e.g. `to_int` -> `int`)
2. make all type coercion functions infallible, return sensible defaults
3. move all type coercion functions into their own "type coercion" section in the docs
4. add missing fallible `parse_*` functions to convert string to type and error on failure (e.g. `parse_int`)

---

I was taking a look at the [example program in our reference documentation](https://master.vector.dev/docs/reference/remap/#example):

```rust
. = parse_json(.message)
.bytes = to_int(.bytes)
.duration = parse_duration(.duration, "s")
.status = to_int(.status)
.timestamp = parse_timestamp(.timestamp)
```

Once we require [error-handling at compile-time](https://github.com/timberio/vector/pull/5941), this program no longer compiles.

I discussed this with @binarylogic. Initially, his proposal was to abort on any error:

```rust
. = parse_json!(.message)
.bytes = to_int!(.bytes)
.duration = parse_duration!(.duration, "s")
.status = to_int!(.status)
.timestamp = parse_timestamp!(.timestamp)
```

But this doesn't capture situations where you want to be more fault tolerant.

An alternative variant would be:

```rust
. = parse_json!(.message)
.bytes = to_int(.bytes) ?? 0
.duration  = parse_duration(.duration, "s") ?? null
.status = to_int(.status) ?? null
.timestamp = parse_timestamp(.timestamp) ?? now()
```

This only requires `.message` to be parsed as JSON, all other mappings fall back to defaults.

When discussing this, we noted that its annoying you have to constantly catch potential errors when coercing values. It would be nicer if all `to_*` coercion functions were infallible. They would default to whatever we consider to be a good default. For example, for a datetime value it would default to "now", but all others would default to their "zero value" (`0`, `false`, `""`, etc).

```rust
. = parse_json!(.message)
.bytes = to_int(.bytes)
.duration  = parse_duration(.duration, "s") ?? null
.status = to_int(.status)
.timestamp = to_timestamp(.timestamp)
```

However, this changes `.status` from `null` to `0` if it can't be coerced, which in this case isn't what we want.

To mitigate this, we add all currently missing `parse_*` variants, which basically take _part of_ the role of the old `to_*` implementation, in that they take a string, and try to parse it to the relevant type, returning an error if they fail.

Additionally, since type coercion is ubiquitous, and now that the function is infallible, we decided to remove the `to_` part of the function:

```rust
. = parse_json!(.message)
.bytes = int(.bytes)
.duration  = parse_duration(.duration, "s") ?? null
.status = parse_int(.status) ?? null
.timestamp = timestamp(.timestamp)
```

This also allows other non-type-coercion `to_` functions to remain fallible, such as `to_syslog_level`. We'll lump all the type-coercion functions together in a "type coercion" section on the docs page, to make them easy to find.

**Comment by binarylogic** [comment]
@JeanMertz and I chatted about this and we've decided against this proposal. Specifically:

> Rename all type coercion to_* functions by removing the to_ part (e.g. to_int -> int)

We can still do this if we'd like, but I don't think it's as necessary given the below conclusions. Also, see https://github.com/timberio/vector/pull/5972#issuecomment-758785452.

> make all type coercion functions infallible, return sensible defaults

This is where we've changed our minds. Given this event:

```json
{
	"status": "not a status"
}
```

We think that this is much more obvious:

```js
.status = to_int(.status) ?? 200
```

Than this:

```js
.status = to_int(.status)
```

1. First, I can set my own default.
2. It's clear to readers that the default is 200.

> move all type coercion functions into their own "type coercion" section in the docs

I'm open to this if we decide to rename the functions.

> add missing fallible parse_* functions to convert string to type and error on failure (e.g. parse_int)

We will still be doing this, but I will open a separate issue. This is still necessary when I want to work exclusively with strings. For example, `to_int(false)` does not fail, but `parse_int(false)` does.

**Comment by binarylogic** [comment]
For clarity, here's the version of the script we settled on:

```js
. = parse_json!(.message)
.bytes = to_int(.bytes) ?? 0
.duration  = parse_duration(.duration, "s") ?? null
.status = to_int(.status) ?? null
.timestamp = parse_timestamp(.timestamp) ?? now()
```
